cryptography/src/bls12381/dkg.rs

```diff
@@ -0,0 +1,1702 @@
+//!     player_shares.insert(player_pk, share);
+//! }
+//!
+//! // Step 5: Observer can also compute the public output
+//! let observer_output = observe::<MinSig, ed25519::PublicKey>(
+//!     round_info,
+//!     dealer_logs,
+//!     1 // Increase this for parallelism.
+//! )?;
+//! println!("DKG completed with threshold {}", observer_output.quorum());
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! For a complete production example with resharing, see [commonware-reshare](https://docs.rs/commonware-reshare).
+use super::primitives::group::Share;
+use crate::{
+    bls12381::primitives::{
+        group::{Element, Scalar},
+        ops::msm_interpolate,
+        poly::{self, new_with_constant, Eval, Poly, Public, Weight},
+        variant::Variant,
+    },
+    transcript::{Summary, Transcript},
+    Digest, PublicKey, Signer,
+};
+use commonware_codec::{Encode, EncodeSize, RangeCfg, Read, ReadExt, Write};
+use commonware_utils::{
+    max_faults, quorum,
+    set::{Ordered, OrderedAssociated},
+    NZU32,
+};
+use core::num::NonZeroU32;
+use rand_core::CryptoRngCore;
+use rayon::{
+    iter::{IntoParallelIterator, ParallelIterator as _},
+    ThreadPoolBuilder,
+};
+use std::collections::BTreeMap;
+use thiserror::Error;
+
+const NAMESPACE: &[u8] = b"commonware-bls12381-dkg";
+
+/// The error type for the DKG protocol.
+///
+/// The only error which can happen through no fault of your own is
+/// [`Error::DkgFailed`]. Everything else only happens if you use a configuration
+/// for [`Info`] or [`Dealer`] which is invalid in some way.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("missing dealer's share from the previous round")]
+    MissingDealerShare,
+    #[error("player is not present in the list of players")]
+    UnknownPlayer,
+    #[error("dealer is not present in the previous list of players")]
+    UnknownDealer(String),
+    #[error("not enough dealers: {0}")]
+    InsufficientDealers(usize),
+    #[error("not enough players: {0}")]
+    InsufficientPlayers(usize),
+    #[error("dkg failed for some reason")]
+    DkgFailed,
+}
+
+/// Recover public polynomial by interpolating coefficient-wise all
+/// polynomials using precomputed Barycentric Weights.
+///
+/// It is assumed that the required number of commitments are provided.
+fn recover_public_with_weights<V: Variant>(
+    commitments: &BTreeMap<u32, poly::Public<V>>,
+    weights: &BTreeMap<u32, poly::Weight>,
+    threshold: u32,
+    concurrency: usize,
+) -> poly::Public<V> {
+    let work = |coeff| {
+        // Extract evaluations for this coefficient from all commitments
+        let evals = commitments
+            .iter()
+            .map(|(dealer, commitment)| poly::Eval {
+                index: *dealer,
+                value: commitment.get(coeff),
+            })
+            .collect::<Vec<_>>();
+
+        // Use precomputed weights for interpolation
+        msm_interpolate(weights, &evals).expect("interpolation should not fail")
```

Comment:
I considered adding support for concurrent MSM (I think `blst` has a template) but given the size of things, I found parallelizing `k` instances of MSM to be faster.